keps/sig-instrumentation/20190605-metrics-validation-and-verification.md

```diff
@@ -0,0 +1,180 @@
+  - [Performance](#performance)
+- [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Test Plan](#test-plan)
+- [Unresolved questions](#unresolved-questions)
+  - [Should changes in histogram buckets/summary objectives be a breaking change?](#should-changes-in-histogram-bucketssummary-objectives-be-a-breaking-change)
+  - [Should code snippets used for unit testing be kept in separate files instead of strings?](#should-code-snippets-used-for-unit-testing-be-kept-in-separate-files-instead-of-strings)
+<!-- /toc -->
+
+## Summary
+
+ This Kubernetes Enhancement Proposal (KEP) builds off of the framework proposed
+in the [Kubernetes Control-Plane Metrics Stability
+KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/20190404-kubernetes-control-plane-metrics-stability.md)
+and proposes a strategy for ensuring conformance of metrics with official
+stability guarantees.
+
+## Motivation
+
+ While the [Kubernetes Control Plane metrics stability KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/20190404-kubernetes-control-plane-metrics-stability.md) provides a framework to define stability levels for control-plane metrics, it does not provide a strategy for verifying and validating conformance to stated guarantees. This KEP intends to propose a framework for validating and verifying metric guarantees.
+
+### Goals
+
+* Given a stable metric, validate that we cannot remove or modify it (other than adding deprecation information).
+* Given a deprecated but stable metric, validate that we cannot remove or modify it until the deprecation period has elapsed.
+* Given an alpha metric which is promoted to be 'stable', automatically include proper instrumentation reviewers (for schema validation and conformance to metrics guidelines).
+
+### Non-Goals
+
+* Conformance testing will not apply to alpha metrics, since alpha metrics do not have stability guarantees.
+
+## Proposal
+
+We will provide validation for metrics under the [new framework](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/20190404-kubernetes-control-plane-metrics-stability.md) with static analysis.
+
+## Design Details
+
+Metrics conformance testing will work in a similar (but not identical) fashion to the generic Kubernetes conformance tests.  Sig-instrumentation will own a directory under `test/instrumentation`. There will be a subdirectory `testdata` in which a file `stable-metrics-list.txt` will live.  This file will be owned by sig-instrumentation. Metrics conformance tests will involve a static analysis script which will traverse the entire codebase and look for metrics which are annotated as 'stable'. For each stable metrics, this script will generate a stringified version of metric metadata (i.e. name, type, labels) which will then be appended together in lexographic order. This will be the output of this script.
```

Comment:
I'm not sure "Conformance testing" is a good label for this, as that has special meaning in kubernetes, and I'm unsure that's what we really want here.

Reply:
Changed to "Stable metrics testing"